programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+    if(item:0=='name_tag' && item:2:'display',
+        global_waystones:pos:'icon':2 = item:2;
+        global_waystones:pos:'name' = _get_name_from_nbt(item:2);
+        slot = if(hand=='mainhand', p~'selected_slot', -1);
+        inventory_set(p, slot, item:1 - 1);
+        sound( 'block.enchantment_table.use', pos );
+        _app_message(p, 'Waypoint named '+global_waystones:pos:'name');
+    true, false);
+);
+
+_mark_player_waypoint(stone_pos, point_pos, uuid) -> (
+    if(!global_waypoints:uuid, global_waypoints:uuid = {}); 
+    global_waypoints:uuid:stone_pos = map(point_pos, floor(_));
+);
+
+// op's waystone screen
+open_waystones_screen() -> (
+    p = player();
+    icons = map(pairs(global_waystones), [_:0, _:1, _:0 + [1,0,0]] );
+    if(length(icons) > 0,  
+        _print_icons_to_screen(_create_warps_screen(p,'fb All Waystones'), icons);
+    ,
+        _app_message(p, 'No waystones to display');
+    );
+);
+
+// open when player clicks on a lodestone
+_open_waypoint_screen(p, pos, uuid) -> (
+    icons = [];
+    bad_keys = [];
+
+    // merge waypoints and waystones data into icons.
+    for(pairs(global_waypoints:uuid),
+        if( global_waystones:(_:0), 
+            icons += ([_:0,global_waystones:(_:0),_:1]);
+        , 
+            bad_keys += _:0;
+        );
+    );
+
+    // Clear player waypoints with missing waystone entires
+    for(bad_keys, delete(global_waypoints:uuid:_));
+
+    // filter out entires in different dimensions.
+    if(!global_settings:'dimensional_crossing', 
+        dimension = current_dimension();
+        icons = filter(icons, _:1:'dimension' == dimension);
+    );
+
+    if(length(icons) > 0,  
+        _print_icons_to_screen(_create_warps_screen(p, str('fb Visited Waypoints', p~'name')), icons);
+    );
+);
+
+_create_warps_screen(p, title) -> (
+    create_screen(p, 'generic_9x6', format(title), _(screen, p, action, data) -> (
+        if(data:'slot' > 53, return());
+        if(action=='pickup',
+            btn = inventory_get(screen, data:'slot');
+            if(!btn, return('cancel'));
+            btn = parse_nbt(btn:2);
+            _warp_player(p, btn:'pos', btn:'dimension', btn:'point');
+            close_screen(screen);
+        );
+        return('cancel');
+    ));
+);
+
+_print_icons_to_screen(screen, icons) -> (
+    for(slice(sort_key(icons, _:1:'name'),0, min(length(icons),53)),
+        stone = _:1;
+        nbt = nbt({
+            'pos'->_:2,
+            'point'->_:0,
+            'dimension'->stone:'dimension' || 'overworld',
+            'display'->{
+                'Lore'->[escape_nbt(str('"%d, %d, %d"', _:0))],
+                'Name'->escape_nbt( str('"%s"', stone:'name'))
+            }
+        });
+        try(
+            if(stone:'icon'=='air', throw('Dont set inventory slot to air'));
+            inventory_set(screen, _i, 1, stone:'icon', nbt);
+        , 'exception', 
+            inventory_set(screen, _i, 1, 'lodestone', nbt);
+        );
```

Comment:
Can't this be replaced with an `if`, without an exception in the middle? Or is there some other thing it's catching there?

Reply:
the Icon is a block, and inventory_set expects an item. they don't always match up 100% of the time. when inventory set fails it throws an error. for example if you place the stone above lava, it would complain about item not found. when you pass it air tho it does not throw an error, but it fails down the line when you try to render air with lore and a name. 

I looked at ways to do this without a try fail block, and this seemed to be the easiest. It would be nice if blocks like lava mapped to a lava bucket but i don't really want to implement that. It would also be nice if you had a way to use items like a bone for the icon, but i have not found a nice way to add that feature. If i come back to this with improvements that will be one of them.